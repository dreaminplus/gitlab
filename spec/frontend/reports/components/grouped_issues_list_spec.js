import { shallowMount } from '@vue/test-utils';
import GroupedIssuesList from '~/reports/components/grouped_issues_list.vue';
import ReportItem from '~/reports/components/report_item.vue';
import SmartVirtualList from '~/vue_shared/components/smart_virtual_list.vue';

describe('Grouped Issues List', () => {
  let wrapper;

  const getPropsData = (resolvedIssues, unresolvedIssues) => ({
    groups: [
      {
        name: 'resolved',
        heading: 'Resolved Heading Text',
        issues: resolvedIssues,
      },
      {
        name: 'unresolved',
        heading: 'Unresolved Heading Text',
        issues: unresolvedIssues,
      },
    ],
  });

  const createComponent = ({ resolvedIssues = [], unresolvedIssues = [], stubs = {} } = {}) => {
    wrapper = shallowMount(GroupedIssuesList, {
      propsData: getPropsData(resolvedIssues, unresolvedIssues),
      stubs,
    });
  };

  const findHeading = groupName => wrapper.find(`[data-testid="${groupName}"`);

  afterEach(() => {
    wrapper.destroy();
    wrapper = null;
  });

  it('renders a smart virtual list with the correct props', () => {
    createComponent({
      resolvedIssues: [{ name: 'foo' }],
      unresolvedIssues: [{ name: 'bar' }],
      stubs: {
        SmartVirtualList,
      },
    });

    expect(wrapper.find(SmartVirtualList).props()).toMatchSnapshot();
  });

  describe('without data', () => {
    beforeEach(createComponent);

    it.each(['unresolved', 'resolved'])('does not a render a header for %s issues', issueName => {
      expect(findHeading(issueName).exists()).toBe(false);
    });

    it.each('resolved', 'unresolved')('does not render report items for %s issues', () => {
      expect(wrapper.contains(ReportItem)).toBe(false);
    });
  });

  describe('with data', () => {
    it.each`
      givenIssues                                | groupName       | expectedHeadingText
      ${{ resolvedIssues: [{ name: 'foo' }] }}   | ${'resolved'}   | ${getPropsData().groups[0].heading}
      ${{ unresolvedIssues: [{ name: 'bar' }] }} | ${'unresolved'} | ${getPropsData().groups[1].heading}
    `(
      'renders the correct heading for $groupName issues',
      ({ givenIssues, groupName, expectedHeadingText }) => {
        createComponent(givenIssues);

        expect(findHeading(groupName).text()).toContain(expectedHeadingText);
      },
    );

    it.each(['resolved', 'unresolved'])('renders all %s issues', issueName => {
      const issues = [{ name: 'foo' }, { name: 'bar' }];

      createComponent({
        [`${issueName}Issues`]: issues,
      });

      expect(wrapper.findAll(ReportItem)).toHaveLength(issues.length);
    });

    it('renders a report item with the correct props', () => {
      createComponent({
        resolvedIssues: [{ name: 'foo' }],
        stubs: {
          ReportItem,
        },
      });

      expect(wrapper.find(ReportItem).props()).toMatchSnapshot();
    });
  });
});
