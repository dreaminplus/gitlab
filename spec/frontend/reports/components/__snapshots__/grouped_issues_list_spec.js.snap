// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grouped Issues List renders a smart virtual list with the correct props 1`] = `
Object {
  "length": 4,
  "remain": 20,
  "rtag": "div",
  "size": 32,
  "wclass": null,
  "wtag": "div",
}
`;

exports[`Grouped Issues List with data renders a report item with the correct props 1`] = `
Object {
  "component": "",
  "isNew": false,
  "issue": Object {
    "name": "foo",
  },
  "showReportSectionStatusIcon": false,
  "status": "none",
  "statusIconSize": 24,
}
`;
